### ISSUE TITLE
 Incorrect Behavior in Updating Cell Value

### ISSUE THREAD

**Issue by theshaurya** [rust-lang/rust#114939]
I tried this code:

```rust
use std::cell::Cell;

const A: Cell<i64> = Cell::new(0);

fn main() {
    A.set(2);
    println!("Value of A: {}", A.get());
}

```

The code should update the value of the Cell to 2 and print "Value of A: 2" or must show compilation error.

Instead, The code still prints "Value of A: 0" instead of "Value of A: 2".

![image](https://github.com/rust-lang/rust/assets/126942649/e80409f5-da83-472e-b4a9-68c71888425d)

**Comment by nbdd0121** [comment]
Run with clippy gives a lint:
```
warning: a `const` item should never be interior mutable
 --> src/main.rs:3:1
  |
3 | const A: Cell<i64> = Cell::new(0);
  | -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | make this a static item (maybe with lazy_static)
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const
  = note: `#[warn(clippy::declare_interior_mutable_const)]` on by default

warning: a `const` item with interior mutability should not be borrowed
 --> src/main.rs:6:5
  |
6 |     A.set(2);
  |     ^
  |
  = help: assign this const to a local or static variable, and use the variable here
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrow_interior_mutable_const
  = note: `#[warn(clippy::borrow_interior_mutable_const)]` on by default

warning: a `const` item with interior mutability should not be borrowed
 --> src/main.rs:7:32
  |
7 |     println!("Value of A: {}", A.get());
  |                                ^
  |
  = help: assign this const to a local or static variable, and use the variable here
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrow_interior_mutable_const
```
